File: Ejercicios/Abstract_factory.java
package Ejercicios;

interface Education {
    String details();
}

interface Experience {
    String details();
}

interface Skills {
    String details();
}

interface Certifications {
    String details();
}


class StudentEducation implements Education {
    public String details() { return "Ingeniería en Software - Itsqmet"; }
}

class StudentExperience implements Experience {
    public String details() { return "Pasantías en empresa de análisis de datos"; }
}

class StudentSkills implements Skills {
    public String details() { return "Python, SQL"; }
}

class StudentCertifications implements Certifications {
    public String details() { return "Curso de Machine Learning"; }
}


class ProfessionalEducation implements Education {
    public String details() { return "Ingeniería en Software - Itsqmet"; }
}

class ProfessionalExperience implements Experience {
    public String details() { return "5 años como desarrollador en BMI"; }
}

class ProfessionalSkills implements Skills {
    public String details() { return "Python, Java, SQL, Gestión de Proyectos"; }
}

class ProfessionalCertifications implements Certifications {
    public String details() { return "AWS Certified Solutions Architect"; }
}


interface CVFactory {
    Education createEducation();
    Experience createExperience();
    Skills createSkills();
    Certifications createCertifications();
}


class StudentCVFactory implements CVFactory {
    public Education createEducation() { return new StudentEducation(); }
    public Experience createExperience() { return new StudentExperience(); }
    public Skills createSkills() { return new StudentSkills(); }
    public Certifications createCertifications() { return new StudentCertifications(); }
}

class ProfessionalCVFactory implements CVFactory {
    public Education createEducation() { return new ProfessionalEducation(); }
    public Experience createExperience() { return new ProfessionalExperience(); }
    public Skills createSkills() { return new ProfessionalSkills(); }
    public Certifications createCertifications() { return new ProfessionalCertifications(); }
}


class CV {
    private String name;
    private Education education;
    private Experience experience;
    private Skills skills;
    private Certifications certifications;

    public CV(CVFactory factory, String name) {
        this.name = name;
        this.education = factory.createEducation();
        this.experience = factory.createExperience();
        this.skills = factory.createSkills();
        this.certifications = factory.createCertifications();
    }

    public void showCV() {
        System.out.println("CV de " + name);
        System.out.println("Educación: " + education.details());
        System.out.println("Experiencia: " + experience.details());
        System.out.println("Habilidades: " + skills.details());
        System.out.println("Certificaciones: " + certifications.details());
    }
}


public class Abstract_factory {
    public static void main(String[] args) {
        String cvType = "Student"; 
        CVFactory factory;

        if (cvType.equals("Student")) {
            factory = new StudentCVFactory();
        } else {
            factory = new ProfessionalCVFactory();
        }

        CV cv = new CV(factory, "Gabriel Proaño");
        cv.showCV();
    }
}

File: Creacional/abstract_factory.py
from abc import ABC, abstractmethod


class Education(ABC):
    @abstractmethod
    def details(self):
        pass

class Experience(ABC):
    @abstractmethod
    def details(self):
        pass

class Skills(ABC):
    @abstractmethod
    def details(self):
        pass

class Certifications(ABC):
    @abstractmethod
    def details(self):
        pass

class StudentEducation(Education):
    def details(self):
        return "Ingeniería en Software - Itsqmet"

class StudentExperience(Experience):
    def details(self):
        return "Pasantías en empresa de análisis de datos"

class StudentSkills(Skills):
    def details(self):
        return "Python, SQL"

class StudentCertifications(Certifications):
    def details(self):
        return "Curso de Machine Learning"


class ProfessionalEducation(Education):
    def details(self):
        return "Ingeniería en Software - Itsqmet"

class ProfessionalExperience(Experience):
    def details(self):
        return "5 años como desarrollador en BMI"

class ProfessionalSkills(Skills):
    def details(self):
        return "Python, Java, SQL, Gestión de Proyectos"

class ProfessionalCertifications(Certifications):
    def details(self):
        return "AWS Certified Solutions Architect"


class CVFactory(ABC):
    @abstractmethod
    def create_education(self):
        pass

    @abstractmethod
    def create_experience(self):
        pass

    @abstractmethod
    def create_skills(self):
        pass

    @abstractmethod
    def create_certifications(self):
        pass


class StudentCVFactory(CVFactory):
    def create_education(self):
        return StudentEducation()

    def create_experience(self):
        return StudentExperience()

    def create_skills(self):
        return StudentSkills()

    def create_certifications(self):
        return StudentCertifications()

class ProfessionalCVFactory(CVFactory):
    def create_education(self):
        return ProfessionalEducation()

    def create_experience(self):
        return ProfessionalExperience()

    def create_skills(self):
        return ProfessionalSkills()

    def create_certifications(self):
        return ProfessionalCertifications()


class CV:
    def __init__(self, factory: CVFactory, name: str):
        self.name = name
        self.education = factory.create_education()
        self.experience = factory.create_experience()
        self.skills = factory.create_skills()
        self.certifications = factory.create_certifications()

    def show_cv(self):
        print(f"📄 CV de {self.name}")
        print("Educación:", self.education.details())
        print("Experiencia:", self.experience.details())
        print("Habilidades:", self.skills.details())
        print("Certificaciones:", self.certifications.details())


cv_type = "Student"  

if cv_type == "Student":
    factory = StudentCVFactory()
else:
    factory = ProfessionalCVFactory()

cv = CV(factory, "Gabriel Proaño")
cv.show_cv()
